tests/steps/test_base_step.py

```diff
@@ -70,3 +137,401 @@ def step_with_config(config: StepConfig) -> None:
+
+    @step
+    def some_step():
+        pass
+
+    step_1 = some_step()
+    step_2 = some_step()
+
+    assert (
+        step_1._internal_execution_parameters["zenml-step_source"]
+        == step_2._internal_execution_parameters["zenml-step_source"]
+    )
+
+
+def test_step_source_execution_parameter_changes_when_signature_changes():
+    """Tests that modifying the input arguments or outputs of a step
+    function changes the step source execution parameter."""
+
+    @step
+    def some_step(some_argument: int) -> int:
+        pass
+
+    step_1 = some_step()
+
+    @step
+    def some_step(some_argument_with_new_name: int) -> int:
+        pass
+
+    step_2 = some_step()
+
+    assert (
+        step_1._internal_execution_parameters["zenml-step_source"]
+        != step_2._internal_execution_parameters["zenml-step_source"]
+    )
+
+    @step
+    def some_step(some_argument: int) -> str:
+        pass
+
+    step_3 = some_step()
+
+    assert (
+        step_1._internal_execution_parameters["zenml-step_source"]
+        != step_3._internal_execution_parameters["zenml-step_source"]
+    )
+
+
+def test_step_source_execution_parameter_changes_when_function_body_changes():
+    """Tests that modifying the step function code changes the step
+    source execution parameter."""
+
+    @step
+    def some_step():
+        pass
+
+    step_1 = some_step()
+
+    @step
+    def some_step():
+        # this is new
+        pass
+
+    step_2 = some_step()
+
+    assert (
+        step_1._internal_execution_parameters["zenml-step_source"]
+        != step_2._internal_execution_parameters["zenml-step_source"]
+    )
+
+
+def test_materializer_source_execution_parameter_changes_when_materializer_changes():
+    """Tests that changing the step materializer changes the materializer
+    source execution parameter."""
+
+    @step
+    def some_step() -> int:
+        return 1
+
+    class MyCustomMaterializer(BuiltInMaterializer):
+        pass
+
+    step_1 = some_step().with_return_materializers(BuiltInMaterializer)
+    step_2 = some_step().with_return_materializers(MyCustomMaterializer)
+
+    key = "zenml-output_materializer_source"
+    assert (
+        step_1._internal_execution_parameters[key]
+        != step_2._internal_execution_parameters[key]
+    )
+
+
+def test_call_step_with_args(int_step_output, step_with_two_int_inputs):
+    """Test that a step can be called with args."""
+    step_with_two_int_inputs()(int_step_output, int_step_output)
+
+
+def test_call_step_with_kwargs(int_step_output, step_with_two_int_inputs):
+    """Test that a step can be called with kwargs."""
+    step_with_two_int_inputs()(input_1=int_step_output, input_2=int_step_output)
+
+
+def test_call_step_with_args_and_kwargs(
+    int_step_output, step_with_two_int_inputs
+):
+    """Test that a step can be called with a mix of args and kwargs."""
+    step_with_two_int_inputs()(int_step_output, input_2=int_step_output)
```

Comment:
No assertion in these tests. Not sure if that's best practice with pytest. I have a feeling these kinds of tests give a false feeling of security.

Reply:
No idea either, do you have any suggestions how we could improve it? Is there some pytest.no_raise or something similar?

Reply:
I've used the second style / syntax before in tests that I wrote:

https://stackoverflow.com/a/20275035/6649687

It's a bit wordier, but I find it a bit clearer and works better when it actually does fail.

Reply:
Actually I quite like the concept of the second link, but I would use it without the parametrize:

```python
from contextlib import ExitStack as does_not_raise

def test_something():
    with does_not_raise():
        do_something()
```

What do you think about that? Maybe also @htahir1 @bcdurak

Reply:
Ok yes I like that better too.

Reply:
I added it for all step and pipeline tests

Reply:
I like it too